_topics/en/soft-fork-activation.md

```diff
@@ -0,0 +1,438 @@
+  illustrated another one of the weaknesses of the ISM mechanism using
+  incrementing version numbers---a miner who signals readiness for
+  the latest proposal (version *n*) also implicitly signals readiness
+  for all previous proposals (e.g. version *n-1*).  This limited the
+  ability to use ISM coordinate multiple upgrades in parallel.
+
+  However, despite the problems at the end of BIP66's activation, ISM
+  was again used to avoid delaying BIP65 further.  There were no
+  activation problems this time.
+
+  #### [2016] BIP9 versionbits: the BIP68/112/113 relative locktime activation
+
+  Several flaws in the ISM mechanism had become apparent by this point
+  and so the new [BIP9][] versionbits mechanism was proposed to fix them:
+
+  - **Unnecessary penalizing miners:** when ISM activated and the
+    minimum block version was incremented, any miners who failed to
+    increment their version would produce invalid blocks even if those
+    blocks didn't violate any of the soft fork's other rules.  For
+    example, in the block that triggered the 4 July 2015 chain split,
+    all the transactions followed the BIP66 strict DER rules---the
+    only reason that block and the subsequent five blocks were rejected,
+    causing over $50,000 in losses, were that the miner produced a
+    version 2 block instead of a version 3 block.
+
+  - **Impractical parallelization:** with ISM, developers felt they
+    needed to wait for one fork to activate before a second fork could
+    be started.
+
+  - **Inability to fail:** ISM attempts didn't have expiration dates.
+    Once node software was released that waited for the activation
+    signal, all nodes from that point needed to continue monitoring for
+    that signal until activation occurred.  There was no way to decide
+    the soft fork was no longer needed.
+
+  - **Loose measurements and unpredictable activation times:** ISM
+    activated whenever 950/1000 blocks signaled readiness.  However,
+    because Bitcoin block production is stochastic, a lucky set of
+    miners with much less than 95% of the total network hashrate could
+    sometimes produce 950 of the last 1,000 blocks, leading to premature
+    activation.  Additionally, not knowing the exact activation time in
+    advance meant it was difficult for protocol developers, merchant
+    system administrators, and mining pool operators to be available in
+    the hours shortly after activation occurred in case there were any
+    problems that needed a quick response.
+
+  BIP9 versionbits attempted to solve these problems by using the block
+  header version field as a bitfield.  Bits in the field were used for
+  signaling only---never to reject a block as invalid---and they could
+  be set in parallel.  Measurements were taken only once every 2,016
+  blocks, minimizing the window during which a small percentage of
+  hashrate would need to be lucky to project a false impression of 95%
```

Comment:
You have a weighted coin that's expected to land on heads 80% of the time, tails 20% of the time.  Imagine two scenarios:

- You flip that coin once every 10 minutes; if after any flip, you saw 950 heads out of the last 1,000, you win.

- Once every 1,000 blocks, you flip that coin 1,000 times.  If you saw 950 or more heads, you win.

In which case is it more probable that you'll win?

Reply:
So I ran some quick simulations on this point and the effect was much smaller than I thought other people had claimed.  Even assuming crazy long deployments of 10 years (!), I couldn't get a fork to activate a 95% threshold with less than about 90% hashrate actually signaling support.  I personally don't think the difference between 90% and 95% is a big deal, so I'm going to edit out this general point from the article.  Thanks for making me question my assumptions!

```python3
import random

weight = 0.90

tries = 0
while True:
    tries += 1

    signals = 0
    for i in range(1000):
        if random.random() < weight:
          signals += 1
    if signals >= 950:
        print(tries)
        break

    if tries > 52000 * 10:
        print("failed")
        break
```

Reply:
I think makes a bigger difference if the window is much shorter (like a day or two instead of a week or two). But otherwise I agree -- I've done the same maths repeatedly and got the same result and found people surprised by it. (So it could be worth discussing it just to correct people's misapprehension)